pycc/ccwfn.py

```diff
@@ -555,7 +590,91 @@ def r_T2(self, o, v, F, ERI, L, t1, t2, Fae, Fme, Fmi, Wmnij, Wmbej, Wmbje, Zmbi
         r_T2 = r_T2 + r_T2.swapaxes(0,1).swapaxes(2,3)
         return r_T2
 
+    # Intermedeates needed for CC3
+    def build_cc3_Wmnij(self, o, v, ERI, t1):
+        contract = self.contract
+        if isinstance(t1, torch.Tensor):
+            W = ERI[o,o,o,o].clone()
+        else:
+            W = ERI[o,o,o,o].copy()
+        tmp = contract('ijma,na->ijmn', ERI[o,o,o,v], t1)
+        W = W + tmp + tmp.swapaxes(0,1).swapaxes(2,3)
+        tmp = contract('ia,mnaf->mnif', t1, ERI[o,o,v,v])
+        W = W + contract('mnif,jf->mnij', tmp, t1)
+        return W
+
+    def build_cc3_Wmbij(self, o, v, ERI, t1, Wmnij):
+        contract = self.contract
+        if isinstance(t1, torch.Tensor):
+            W = ERI[o,v,o,o].clone()
+        else:
+            W = ERI[o,v,o,o].copy()
+        W = W - contract('mnij,nb->mbij', Wmnij, t1)
+        W = W + contract('mbie,je->mbij', ERI[o,v,o,v], t1)
+        tmp = ERI[o,v,v,o] + contract('mbef,jf->mbej', ERI[o,v,v,v], t1)
+        W = W + contract('ie,mbej->mbij', t1, tmp)
+        return W
+
+    def build_cc3_Wmnie(self, o, v, ERI, t1):
+        contract = self.contract
+        if isinstance(t1, torch.Tensor):
+            W = ERI[o,o,o,v].clone()
+        else:
+            W = ERI[o,o,o,v].copy()
+        W = W + contract('if,mnfe->mnie', t1, ERI[o,o,v,v])
+        return W
 
+    def build_cc3_Wamef(self, o, v, ERI, t1):
+        contract = self.contract
+        if isinstance(t1, torch.Tensor):
+            W = ERI[v,o,v,v].clone()
+        else:
+            W = ERI[v,o,v,v].copy()
+        W = W - contract('na,nmef->amef', t1, ERI[o,o,v,v])
+        return W   
+
+    def build_cc3_Wabei(self, o, v, ERI, t1):
```

Comment:
I'm not going to request changes to this function, but I do think we should re-visit it at some point.  I think the structure you've adopted is intended to follow that from the corresponding functions in Psi4.  However, many of the algorithmic issues that are important in that C++ code don't really apply to a Python variant, and so we might be able to simply this section and make it more readable.

Reply:
That makes sense. I will keep it in mind and think about it.